cpp/tests/test_abm_serialization.cpp

```diff
@@ -0,0 +1,298 @@
+#include "abm/infection_state.h"
+#include "abm/parameters.h"
+#include "abm/test_type.h"
+#include "abm/testing_strategy.h"
+#include "abm/vaccine.h"
+#include "matchers.h"
+#include "memilio/epidemiology/age_group.h"
+#include "memilio/io/json_serializer.h"
+#include "memilio/utils/custom_index_array.h"
+#include "memilio/utils/uncertain_value.h"
+#include "models/abm/location.h"
+#include "models/abm/person.h"
+#include "models/abm/time.h"
+#include "models/abm/trip_list.h"
+#include "models/abm/model.h"
+
+#ifdef MEMILIO_HAS_JSONCPP
+
+void test_equal_json_representation(const Json::Value& test_json, const Json::Value& reference_json)
+{
+    // write the resulting json value and the reference value to string to compare their representations.
+    Json::StreamWriterBuilder swb;
+    swb["indentation"] = " ";
+    auto js_writer     = std::unique_ptr<Json::StreamWriter>(swb.newStreamWriter());
+    std::stringstream test_str, reference_str;
+    js_writer->write(reference_json, &reference_str);
+    js_writer->write(test_json, &test_str);
+    // we compare strings here, as e.g. Json::Int(5) != Json::Uint(5), but their json representation is the same
+    EXPECT_EQ(test_str.str(), reference_str.str());
+}
+
+/**
+ * @brief Test de- and serialization of an object by comparing its json representation.
+ *
+ * Test that a json value x representing type T is equal to serialize(deserialize(x)) w.r.t json representation.
+ *
+ * Assuming the (de)serialization functions' general behavior is independent of specific values of member variables,
+ * i.e. the function does not contain conditionals (`if (t > 0)`), optionals (`add_optional`/`expect_optional`), etc.,
+ * and assuming that deserialize is injective (meaning that two unequal instances of T do not have the same json
+ * representation, which can happen e.g. if not all member variables are serialized),
+ * this sufficiently tests that serialize and deserialize are inverse functions to each other.
+ *
+ * @tparam T The type to test.
+ * @param reference_json A json value representing an instance of T.
+ */
+template <class T>
+void test_json_serialization_by_representation(const Json::Value& reference_json)
+{
+    // check that the json is deserializable (i.e. a valid representation)
+    auto t_result = mio::deserialize_json(reference_json, mio::Tag<T>());
+    ASSERT_THAT(print_wrap(t_result), IsSuccess());
+
+    // check that the resulting type T is serializable
+    auto json_result = mio::serialize_json(t_result.value());
+    ASSERT_THAT(print_wrap(json_result), IsSuccess());
+
+    test_equal_json_representation(json_result.value(), reference_json);
```

Comment:
Yes, the serialization is predictable, but comparing json values directly means that the reference values must be exactly right. For one, this makes writing these tests a little more cumbersome, so I could change it. But It also gives better results IMO, since GTEST has a built in diff view for string comparisons.

Reply:
I think we want to be as strict as possible with types, since other serializers (e.g. binary) are more strict than json. 

The solution for nicer output is to put the conversion to string you wrote in the comparison function into a custom print function similar to print_wrap for Eigen matrix.

Reply:
The default print (gtest seems to have one for json values) actually looks better for values, but lacks a diff view, so you have to manually search two json representations for a difference - if there is one.
Say we serialize a 0 twice using Json::Int and Json::UInt, respectively. These are not equal as Json::Value, but gtest prints the exact same representation twice. And I don't know how to find out which type is mismatched in larger Values. I doubt that it is worth the effort to find out and check.

Reply:
I see, I haven't seen the diff view before. Only happens with multi line strings. 

I think the types are important. I think it's worth to write a custom matcher for this. For simplicity, something like this. If the json is an object or list, iterate over the members. otherwise compare with normal equality `a == b`. For slightly more info, compare `a.type() == b.type()` before doing `a == b`. Then print the two full jsons, the path to the difference and the sub tree (probably a single value) that is different using the print function you already written.

Reply:
I added a matcher that recursively checks for type, then value equality. It uses depth first search for objects or arrays and prints the first error found.
I could not get the print wrap to work as intended, it seemed like it was used only partially (some indents used tabs instead of spaces ans some object names were missing). But the same function as a PrintTo works fine.